packages/components/src/components/select/select/select.tsx

```diff
@@ -0,0 +1,393 @@
+
+  /** The selected value. */
+  @Prop({ mutable: true }) public value?: string;
+
+  /** The validation state. */
+  @Prop() public state?: SelectState = 'none';
+
+  /** The message styled depending on validation state. */
+  @Prop() public message?: string = '';
+
+  /** Show or hide label. For better accessibility it is recommended to show the label. */
+  @Prop() public hideLabel?: BreakpointCustomizable<boolean> = false;
+
+  /** Disables the select */
+  @Prop() public disabled?: boolean = false;
+
+  /** A Boolean attribute indicating that an option with a non-empty string value must be selected. */
+  @Prop() public required?: boolean = false;
+
+  /** Changes the direction to which the dropdown list appears. */
+  @Prop() public dropdownDirection?: SelectDropdownDirection = 'auto';
+
+  /** Adapts the select color depending on the theme. */
+  @Prop() public theme?: Theme = 'light';
+
+  /** Emitted when the selection is changed. */
+  @Event({ bubbles: false }) public update: EventEmitter<SelectUpdateEventDetail>;
+
+  @State() private isOpen = false;
+  @State() private srHighlightedOptionText = '';
+
+  private nativeSelect: HTMLSelectElement;
+  private comboboxContainer: HTMLDivElement;
+  private combobox: HTMLButtonElement;
+  private listElement: HTMLDivElement;
+  private selectOptions: SelectOption[] = [];
+  private form: HTMLFormElement;
+  private isWithinForm: boolean;
+  private preventOptionUpdate = false; // Used to prevent value watcher from updating options when options are already updated
+
+  private searchString: string = '';
+  private searchTimeout: NodeJS.Timeout | number = null;
+
+  @Listen('internalOptionUpdate')
+  public updateOptionHandler(e: Event & { target: SelectOption }): void {
+    e.stopPropagation();
+    this.updateSelectedOption(e.target);
+  }
+
+  @Watch('value')
+  public onValueChange(): void {
+    // When setting initial value the watcher gets called before the options are defined
+    if (this.selectOptions.length > 0) {
+      if (!this.preventOptionUpdate) {
+        updateSelectOptions(this.selectOptions, this.value);
+      }
+      this.preventOptionUpdate = false;
+      if (this.isWithinForm) {
+        updateNativeSelectOption(this.nativeSelect, this.selectOptions);
+      }
+    }
+  }
+
+  public connectedCallback(): void {
+    document.addEventListener('mousedown', this.onClickOutside, true);
+    this.form = getClosestHTMLElement(this.host, 'form');
+    this.isWithinForm = !!this.form;
+  }
+
+  public componentWillLoad(): void {
+    this.updateOptions();
+    updateSelectOptions(this.selectOptions, this.value);
+    if (this.isWithinForm) {
+      this.nativeSelect = initNativeSelect(this.host, this.name, this.disabled, this.required);
+      updateNativeSelectOption(this.nativeSelect, this.selectOptions);
+    }
+  }
+
+  public componentDidLoad(): void {
+    getShadowRootHTMLElement(this.host, 'slot').addEventListener('slotchange', this.onSlotchange);
```

Comment:
Here we do it imperatively, I think in `p-select-option` it was done declaratively in JSX.  
Do both work equally now? Why not stay consistent?